### ISSUE TITLE
Running more than 48 hrs on pl.divide_and_conquer_pipeline with 8 cores

### ISSUE THREAD

**Issue by onionpork** [tanaylab/metacells#16]
I am using a Linux cluster that runs a 64-bit CentOS 7.6 operating system with SGE 8.6.4 job scheduler and it is managed using Ansible.
With the 

- Intel Intel(R) Xeon(R) Silver 4110 CPU @ 2.10GHz (Duel CPU, total 16 cores)

- 192GB of RAM (16GB x 12 DDR4 2666MHz ECC REG Memory Module)

- 4 NVIDIA GPU card (Tesla P100 16GB CoWoS HBM2 PCIe 3.0 -- Passive Cooling)

- 4.6TB Solid State Drive (SSD)

`grep flags /proc/cpuinfo | head -1`
`flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_ppin intel_pt ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke spec_ctrl intel_stibp flush_l1d`

I requested 8 cores for one node with 1T memory, upgraded to metacells 0.7.0, and tested on the pbmc163 dataset you provided. 
Prints:
`set PBMC.var[properly_sampled_gene]: 22637 true (69.15%) out of 32738 bools

set PBMC.var[excluded_gene]: 20 true (0.06109%) out of 32738 bools

set PBMC.var[noisy_lonely_gene]: 0 true (0%) out of 32738 bools

set PBMC.var[clean_gene]: 22617 true (69.08%) out of 32738 bools

set PBMC.obs[properly_sampled_cell]: 149825 true (91.79%) out of 163234 bools

set PBMC.obs[clean_cell]: 149825 true (91.79%) out of 163234 bools

set PBMC.clean.obs[full_cell_index]: 149825 int64s

set PBMC.clean.var[full_gene_index]: 22617 int64s

set PBMC.clean.var[related_genes_module]: 21549 outliers (95.28%) out of 22617 int32 elements with 73 groups with mean size 14.63

set PBMC.clean.varp[related_genes_similarity]: csr_matrix 22617 X 22617 float32s

set PBMC.clean.var[rare_gene_module_0]: 4 true (0.01769%) out of 22617 bools

set PBMC.clean.var[rare_gene_module_1]: 29 true (0.1282%) out of 22617 bools

set PBMC.clean.var[rare_gene]: 33 true (0.1459%) out of 22617 bools

set PBMC.clean.obs[cells_rare_gene_module]: 149102 outliers (99.52%) out of 149825 int32 elements with 2 groups with mean size 361.5

set PBMC.clean.obs[rare_cell]: 723 true (0.4826%) out of 149825 bools`

Above steps are quick, like minutes. But then it stuck there for the next 48 hrs until my request time is over.
Thank you very much for any suggestions.

**Comment by orenbenkiki** [comment]
Sorry for the late reply.

I've had many people successfully running the vignette so this is "very surprising". I just rechecked and computing the metacells for PBMC using the divide-and-conquer method takes less than 20GB of memory so unless your machine is running another heavy process in parallel (e.g., R sessions are notorious memory hogs), running out of memory really shouldn't be the issue.

1. Just to make sure, I recommend you run `top` or `htop` in parallel to the execution and see whether the machine indeed has plenty of unused RAM as expected. It should also show whether the metacells computation is actively consuming CPU cycles at a healthy rate (a few or at least one process consuming close to 100% CPU), is running very slowly (low-non-zero-% CPU usage) or is actually blocked (0% CPU) for some mysterious reason.

2. Ruling out memory usage as the culprit, all I can suggest is the usual trick of re-running the program while adding, immediately after `import metacells as mc`, the following lines:

```python
import logging
mc.ut.setup_logger(level=logging.DEBUG)
```

This should generate a more detailed log file that could may help figure out what is going one.

3. As a last resort, for an even more ridiculously detailed log file you can set up the following environment variables before the run:

```bash
export METACELLS_COLLECT_TIMING=true
export METACELLS_LOG_ALL_STEPS=true
```

This will emit a timestamp before and after every low-level operation of interest and will show whether the program is in an endless loop, is stuck in a specific operation, or is making very slow progress.

**Comment by orenbenkiki** [comment]
I have optimized the rare gene modules detection in the about-to-be-released version 0.9 which may also help. Note that version 0.9 is different from version 0.8 (see the project README). It is in final testing phases, currently you can only install it from the head version in github.

**Comment by orenbenkiki** [comment]
Version 0.9 is now published so closing this as done.
